hermeznetwork/hermez-node
# instrumenting the application

### What does this MR does?

Instrumenting the node with exp vars for Prometheus can collect and store.

- Create the metric package;
- Create a gin middleware to measure the API requests;
- Collect API error responses;
- Create a Prometheus collector to instrument the database connections;
- Create a method to measure the method execution time;
- Measure the proof server request;
- Add block reorg count metric;

### Why we need it?

For we have better node metrics for Prometheus to show inside the Grafana dashboard.

### How to test?

- run the node with the debug API on:
```
$ go run ./cli/node --mode coord --cfg cli/node/cfg.buidler.toml
``` 

- check the metrics `/debug/metrics`;

closes https://github.com/hermeznetwork/hermez-node/issues/628

#### Dashboard sample:
![Screen Shot 2021-03-19 at 01 05 57](https://user-images.githubusercontent.com/2406457/111730084-613b4400-884f-11eb-98ae-33085af0dc98.png)
